test/unit/utils/sql.test.ts

```diff
@@ -0,0 +1,590 @@
+    });
+  });
+
+  it('does not consider the token to be a replacement if it is part of a nested $ quoted string', () => {
+    const sql = injectReplacements(`SELECT * FROM users WHERE id = $tag1$ $tag2$ ? $tag2$ $tag1$`, dialect, [1]);
+
+    expectsql(sql, {
+      default: `SELECT * FROM users WHERE id = $tag1$ $tag2$ ? $tag2$ $tag1$`,
+    });
+  });
+
+  it('does consider the token to be a replacement if it is in between two identifiers that look like $ quoted strings', () => {
+    const sql = injectReplacements(`SELECT z$$ ? x$$ * FROM users`, dialect, [1]);
+
+    expectsql(sql, {
+      default: `SELECT z$$ 1 x$$ * FROM users`,
+    });
+  });
+
+  it('does not consider the token to be a replacement if it is part of a string with a backslash escaped quote', () => {
+    const sql = injectReplacements(`SELECT * FROM users WHERE id = '\\'?' OR id = ?`, dialect, [1]);
+
+    expectsql(sql, {
+      default: `SELECT * FROM users WHERE id = '\\'?' OR id = 1`,
+    });
+  });
+
+  it('considers the token to be a replacement if it is outside a string ending with an escaped backslash', () => {
+    const sql = injectReplacements(`SELECT * FROM users WHERE id = '\\\\' OR id = ?`, dialect, [1]);
+
+    expectsql(sql, {
+      default: `SELECT * FROM users WHERE id = '\\\\' OR id = 1`,
+    });
+  });
+
+  it('does not consider the token to be a replacement if it is part of a string with an escaped backslash followed by a backslash escaped quote', () => {
+    const sql = injectReplacements(`SELECT * FROM users WHERE id = '\\\\\\'?' OR id = ?`, dialect, [1]);
+
+    expectsql(sql, {
+      default: `SELECT * FROM users WHERE id = '\\\\\\'?' OR id = 1`,
+    });
+  });
+
+  it('does not consider the token to be a replacement if it is in a single line comment', () => {
+    const sql = injectReplacements(`
+      SELECT * FROM users -- WHERE id = ?
+      WHERE id = ?
+    `, dialect, [1]);
+
+    expectsql(sql, {
+      default: `
+        SELECT * FROM users -- WHERE id = ?
+        WHERE id = 1
+      `,
+    });
+  });
+
+  it('does not consider the token to be a replacement if it is in string but a previous comment included a string delimiter', () => {
+    const sql = injectReplacements(`
+      SELECT * FROM users -- '
+      WHERE id = ' ? '
+    `, dialect, [1]);
+
+    expectsql(sql, {
+      default: `
+        SELECT * FROM users -- '
+        WHERE id = ' ? '
+      `,
+    });
+  });
+
+  it('does not consider the token to be a replacement if it is in a single line comment', () => {
+    const sql = injectReplacements(`
+      SELECT * FROM users /*
+      WHERE id = ?
+      */
+      WHERE id = ?
+    `, dialect, [1]);
+
+    expectsql(sql, {
+      default: `
+        SELECT * FROM users /*
+        WHERE id = ?
+        */
+        WHERE id = 1
+      `,
+    });
+  });
+
+  // https://github.com/sequelize/sequelize/issues/14358
+  it('does not parse ?& and ?| operators as replacements (#14358)', async () => {
```

Comment:
Do we want to test this one on all dialects even though it's postgres exclusive?

Reply:
Yep! I'd prefer it to be consistent across all dialects, it could be added to other dialects in the future if they start supporting arrays (last I heard arrays were added to the SQL spec)